File: src/ChpokkWeb/SystemFiles/Templates/ProjectTemplates/CSharp/Silverlight/1041/BusinessApplication/Helpers/DataFieldExtensions.cs
namespace $safeprojectname$
{
    using System;
    using System.Windows;
    using System.Windows.Controls;
    using System.Windows.Data;

    /// <summary>
    ///     <see cref="DataField"/> で操作を実行するための拡張メソッドを提供します。
    /// </summary>
    public static class DataFieldExtensions
    {
        /// <summary>
        /// <see cref="DataField" /> の <see cref="TextBox" /> コントロールを別のコントロールに置換し、バインディングを更新します。
        /// </summary>
        /// <param name="field"><see cref="TextBox"/> が置換される <see cref="DataField"/> です。</param>
        /// <param name="newControl"><see cref="DataField.Content" /> として設定する新しいコントロールです。</param>
        /// <param name="dataBindingProperty">データ バインディングに使用されるコントロールのプロパティです。</param>        
        public static void ReplaceTextBox(this DataField field, FrameworkElement newControl, DependencyProperty dataBindingProperty)
        {
            field.ReplaceTextBox(newControl, dataBindingProperty, null);
        }

        /// <summary>
        /// <see cref="DataField" /> の <see cref="TextBox" /> コントロールを別のコントロールに置換し、バインディングを更新します。
        /// </summary>
        /// <param name="field"><see cref="TextBox"/> が置換される <see cref="DataField"/> です。</param>
        /// <param name="newControl"><see cref="DataField.Content" /> として設定する新しいコントロールです。</param>
        /// <param name="dataBindingProperty">データ バインディングに使用されるコントロールのプロパティです。</param>        
        /// <param name="bindingSetupFunction">
        ///  <paramref name="newControl"/> に適用される前に新しく生成されたバインディングでパラメーターの変更に使用できる、オプションの <see cref="Action"/> です
        /// </param>
        public static void ReplaceTextBox(this DataField field, FrameworkElement newControl, DependencyProperty dataBindingProperty, Action<Binding> bindingSetupFunction)
        {
            if (field == null)
            {
                throw new ArgumentNullException("field");
            }

            if (newControl == null)
            {
                throw new ArgumentNullException("newControl");
            }

            // 既存のバインディングをコピーして新しいバインディングを構築し、呼び出し元が実行する変更の bindingSetupFunction に送信します。
            Binding newBinding = field.Content.GetBindingExpression(TextBox.TextProperty).ParentBinding.CreateCopy();

            if (bindingSetupFunction != null)
            {
                bindingSetupFunction(newBinding);
            }

            // 置換フィールド
            newControl.SetBinding(dataBindingProperty, newBinding);
            field.Content = newControl;
        }

        /// <summary>
        /// 別の <see cref="Binding"/> オブジェクトからすべてのプロパティをコピーして、新しい <see cref="Binding"/> オブジェクトを作成します。
        /// </summary>
        /// <param name="binding">プロパティ値のコピー元の <see cref="Binding"/> です</param>
        /// <returns>新しい <see cref="Binding"/> オブジェクトです。</returns>
        private static Binding CreateCopy(this Binding binding)
        {
            if (binding == null)
            {
                throw new ArgumentNullException("binding");
            }

            Binding newBinding = new Binding()
            {
                BindsDirectlyToSource = binding.BindsDirectlyToSource,
                Converter = binding.Converter,
                ConverterParameter = binding.ConverterParameter,
                ConverterCulture = binding.ConverterCulture,
                Mode = binding.Mode,
                NotifyOnValidationError = binding.NotifyOnValidationError,
                Path = binding.Path,
                UpdateSourceTrigger = binding.UpdateSourceTrigger,
                ValidatesOnExceptions = binding.ValidatesOnExceptions
            };

            if (binding.ElementName != null)
            {
                newBinding.ElementName = binding.ElementName;
            }
            else if (binding.RelativeSource != null)
            {
                newBinding.RelativeSource = binding.RelativeSource;
            }
            else
            {
                newBinding.Source = binding.Source;
            }

            return newBinding;
        }
    }
}

File: src/ChpokkWeb/SystemFiles/Templates/ProjectTemplates/VisualBasic/Silverlight/1031/BusinessApplication/Helpers/DataFieldExtensions.vb
Imports System
Imports System.Runtime.CompilerServices
Imports System.Windows
Imports System.Windows.Controls
Imports System.Windows.Data

Public Module DataFieldExtensions
    ''' <summary>
    ''' Ersetzt das <see cref="TextBox" />-Steuerelement eines <see cref="DataField" /> mit einem anderen Steuerelement und aktualisiert die Bindungen.
    ''' </summary>
    ''' <param name="field">Das <see cref="DataField"/>, dessen <see cref="TextBox"/> ersetzt wird.</param>
    ''' <param name="newControl">Das neue Steuerelement, das als <see cref="DataField.Content" /> festgelegt wird.</param>
    ''' <param name="dataBindingProperty">Die Eigenschaft des Steuerelements, die für die Datenbindung verwendet wird.</param>        
    ''' <param name="bindingSetupFunction">
    ''' Eine optionale <c>Action</c>, die verwendet werden kann, um Parameter der neu erzeugten Bindung zu ändern, bevor diese auf <paramref name="newControl"/> angewendet wird.
    ''' </param>
    <Extension()> _
    Public Sub ReplaceTextBox(ByVal field As DataField, ByVal newControl As FrameworkElement, ByVal dataBindingProperty As DependencyProperty, _
                              Optional ByVal bindingSetupFunction As Action(Of Binding) = Nothing)
        If field Is Nothing Then
            Throw New ArgumentNullException("field")
        End If

        If newControl Is Nothing Then
            Throw New ArgumentNullException("newControl")
        End If

        ' Erstellt eine neue Bindung, indem die bestehende Bindung kopiert und ggf. an eine vom Benutzer angegebene Funktion übergeben wird.
        Dim newBinding As Binding = field.Content.GetBindingExpression(TextBox.TextProperty).ParentBinding.CreateCopy()

        If bindingSetupFunction IsNot Nothing Then
            bindingSetupFunction(newBinding)
        End If

        ' Feld ersetzen
        newControl.SetBinding(dataBindingProperty, newBinding)
        field.Content = newControl
    End Sub

    ''' <summary>
    ''' Erstellt ein neues <see cref="Binding"/>-Objekt, indem alle Eigenschaften eines anderen <see cref="Binding"/>-Objekts kopiert werden.
    ''' </summary>
    ''' <param name="binding"><see cref="Binding"/>, von der die Eigenschaftswerte kopiert werden</param>
    ''' <returns>Ein neues <see cref="Binding"/>-Objekt.</returns>
    <Extension()> _
    Public Function CreateCopy(ByVal binding As Binding) As binding
        If binding Is Nothing Then
            Throw New ArgumentNullException("binding")
        End If

        Dim newBinding As New Binding()

        newBinding.BindsDirectlyToSource = binding.BindsDirectlyToSource
        newBinding.Converter = binding.Converter
        newBinding.ConverterParameter = binding.ConverterParameter
        newBinding.ConverterCulture = binding.ConverterCulture
        newBinding.Mode = binding.Mode
        newBinding.NotifyOnValidationError = binding.NotifyOnValidationError
        newBinding.Path = binding.Path
        newBinding.UpdateSourceTrigger = binding.UpdateSourceTrigger
        newBinding.ValidatesOnExceptions = binding.ValidatesOnExceptions

        If binding.ElementName IsNot Nothing Then
            newBinding.ElementName = binding.ElementName
        ElseIf binding.RelativeSource IsNot Nothing Then
            newBinding.RelativeSource = binding.RelativeSource
        Else
            newBinding.Source = binding.Source
        End If

        Return newBinding
    End Function
End Module

